From 788625e7ba18485e52382841e88c757b150d8d99 Mon Sep 17 00:00:00 2001
From: Soren Brinkmann <soren.brinkmann@xilinx.com>
Date: Wed, 5 Jun 2013 13:24:42 -0700
Subject: [PATCH 404/628] net: xemacps: Free net_device after it is used last

git://github.com/Xilinx/linux-xlnx.git xilinx-v14.7
commit 6a3a71448395354284cf8cd664cc9d19d2cce928

The net_device structure is freed to early. Free it after its last user
is done.

Signed-off-by: Soren Brinkmann <soren.brinkmann@xilinx.com>
Signed-off-by: Liming Wang <liming.wang@windriver.com>
---
 drivers/net/ethernet/xilinx/xilinx_emacps.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_emacps.c b/drivers/net/ethernet/xilinx/xilinx_emacps.c
index 030ad5e..076488f 100644
--- a/drivers/net/ethernet/xilinx/xilinx_emacps.c
+++ b/drivers/net/ethernet/xilinx/xilinx_emacps.c
@@ -2785,8 +2785,6 @@ static int __exit xemacps_remove(struct platform_device *pdev)
 		unregister_netdev(ndev);
 		free_irq(ndev->irq, ndev);
 		iounmap(lp->baseaddr);
-		free_netdev(ndev);
-		platform_set_drvdata(pdev, NULL);
 
 		clk_notifier_unregister(lp->devclk, &lp->clk_rate_change_nb);
 		if (!pm_runtime_suspended(&pdev->dev)) {
@@ -2798,6 +2796,8 @@ static int __exit xemacps_remove(struct platform_device *pdev)
 		}
 		clk_put(lp->devclk);
 		clk_put(lp->aperclk);
+
+		free_netdev(ndev);
 	}
 
 	return 0;
-- 
1.7.5.4

